include/deal.II/matrix_free/fe_point_evaluation.h

```diff
@@ -816,15 +852,53 @@ void
 FEPointEvaluation<n_components, dim, spacedim, Number>::reinit(
   const typename Triangulation<dim, spacedim>::cell_iterator &cell,
   const ArrayView<const Point<dim>> &                         unit_points)
+{
+  // If using the fast path, we need to precompute the mapping data.
+  if (!poly.empty())
+    internal::FEPointEvaluation::compute_mapping_data_for_generic_points(
+      *mapping, cell, unit_points, update_flags_mapping, mapping_data);
+
+  // then call the other version of this function with the precomputed data
+  reinit(cell, unit_points, mapping_data);
+}
+
+
+
+template <int n_components, int dim, int spacedim, typename Number>
+void
+FEPointEvaluation<n_components, dim, spacedim, Number>::reinit(
+  const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+  const ArrayView<const Point<dim>> &                         unit_points,
+  const dealii::internal::FEValuesImplementation::MappingRelatedData<dim,
+                                                                     spacedim>
+    &precomputed_mapping_data)
 {
   this->unit_points.resize(unit_points.size());
   std::copy(unit_points.begin(), unit_points.end(), this->unit_points.begin());
 
   if (!poly.empty())
-    fill_mapping_data_for_generic_points(cell,
-                                         unit_points,
-                                         update_flags_mapping,
-                                         mapping_data);
+    {
+      // Check the mapping data for consistency.
+      if (update_flags_mapping & update_jacobians)
+        Assert(precomputed_mapping_data.jacobians.size() == unit_points.size(),
+               ExcDimensionMismatch(precomputed_mapping_data.jacobians.size(),
+                                    unit_points.size()));
+      if (update_flags_mapping & update_inverse_jacobians)
+        Assert(precomputed_mapping_data.inverse_jacobians.size() ==
+                 unit_points.size(),
+               ExcDimensionMismatch(
+                 precomputed_mapping_data.inverse_jacobians.size(),
+                 unit_points.size()));
+
+      if (update_flags_mapping & update_quadrature_points)
+        Assert(precomputed_mapping_data.inverse_jacobians.size() ==
+                 unit_points.size(),
+               ExcDimensionMismatch(
+                 precomputed_mapping_data.quadrature_points.size(),
+                 unit_points.size()));
+
+      mapping_data = precomputed_mapping_data;
```

Comment:
I believe we might, as a possible follow-up work, implement a specific assignment function that avoids re-allocation of the various attributes in `MappingRelatedData` as they happen via `operator=`, i.e., do some copy-of-content and resize if necessary (or do we know if typical STL implementations avoid re-allocation in `operator=`, which is at least technically possible?).